### PR TITLE
Bench and restructure staking

### LINKED ISSUE #975: Staking benchmark

### PR DESCRIPTION
**Pull request by AurevoirXavier**
Closes #975.

### CODE REVIEW

File: pallet/staking/traits/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (boundless-forest) [inline pallet/staking/traits/src/lib.rs:1]: Add license.

File: pallet/staking/src/mock.rs [hunk 2]
```diff
@@ -228,6 +229,8 @@ frame_support::parameter_types! {
 impl darwinia_staking::Config for Runtime {
 	type Deposit = Deposit;
 	type Kton = KtonStaking;
+	#[cfg(feature = "runtime-benchmarks")]
+	type KtonHelper = KtonMinting;
```
Comment (boundless-forest) [inline pallet/staking/src/mock.rs:233]: Why don't use `type Kton = KtonMinting` in the pallet-deposit directly?

File: runtime/pangolin/tests/tests.rs [hunk 3]
```diff
@@ -1,7 +1,10 @@
 pub mod mock;
 
+#[cfg(test)]
```
Comment (boundless-forest) [inline runtime/pangolin/tests/tests.rs:3]: Please sync this changes to other runtimes.

Comment (AurevoirXavier) [reply runtime/pangolin/tests/tests.rs:3]: Sorry, I forgot that this is an external test. Removed.

File: pallet/staking/src/mock.rs (hunk 2 above)
Comment (AurevoirXavier) [reply pallet/staking/src/mock.rs:233]: Because they are different interfaces/traits.

Comment (boundless-forest) [reply pallet/staking/src/mock.rs:233]: What's the difference between them? Seems both implement the `darwinia_deposit::SimpleAsset`.

File: pallet/staking/src/benchmarking.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+			})
+			.collect()
+	}
+
+	#[benchmark]
+	fn stake(x: Linear<0, 255>) {
+		let a = frame_benchmarking::whitelisted_caller();
+
+		T::RingCurrency::make_free_balance_be(&a, 256 * UNIT);
+		T::KtonHelper::mint(&a, UNIT).unwrap();
```
Comment (boundless-forest) [inline pallet/staking/src/benchmarking.rs:56]: Suggest replacing `KtonHelper` with `type Kton = KtonMinting` of the `darwinia-deposit` config. They are the same in your code.

Comment (AurevoirXavier) [reply pallet/staking/src/benchmarking.rs]: Great idea.

File: pallet/staking/src/lib.rs [hunk 5]
```diff
@@ -62,120 +68,26 @@ use sp_runtime::{
...
+	// Deposit helper for runtime benchmark.
+	#[cfg(feature = "runtime-benchmarks")]
+	use darwinia_deposit::Config as DepositConfig;
+	/// Empty trait acts as a place holder to satisfy the `#[pallet::config]` macro.
+	#[cfg(not(feature = "runtime-benchmarks"))]
+	pub trait DepositConfig {}
+
 	#[pallet::config]
-	pub trait Config: frame_system::Config {
+	pub trait Config: frame_system::Config + DepositConfig {
```
Comment (boundless-forest) [inline pallet/staking/src/lib.rs:84]: Do we still need the `DepositConfig` bound after removing the `KtonHelper`?

Comment (AurevoirXavier) [reply pallet/staking/src/lib.rs:84]: Yep. Because of `<T as deposit::Config>::Kton`.

Comment (AurevoirXavier) [reply pallet/staking/src/lib.rs:84]: Also
```rs
fn deposit_for<T>(who: &T::AccountId, count: u32) -> Vec<DepositId<T>>
	where
		T: Config + darwinia_deposit::Config,
{
	(0..count.min(<<T as Config>::MaxDeposits>::get()) as u8)
		.map(|x| {
			<darwinia_deposit::Pallet<T>>::lock(
				RawOrigin::Signed(who.to_owned()).into(),
				UNIT,
				1,
			)
			.unwrap();

			x.into()
		})
		.collect()
}
```

**Review by boundless-forest — APPROVED** [review]
LGTM

**Pull request merged**